keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,899 @@
+
+## Motivation
+
+Being able to take application consistent snapshots/backups is a hard requirement to protect stateful workloads in Kubernetes. It's required for many applications to temporally quiesce themselves before taking snapshots of their PersistentVolume(s) and unquiesce afterwards to ensure application consistency. A mechanism to send a quiesce/unquiesce command to `Pod`(s) is needed.
+
+The first attempt was to introduce an [ExecutionHook](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190120-execution-hook-design.md) CRD which allows arbitrary commands to be executed against containers, and a controller which manages the lifecycle of the custom resources. This approach, though solves the application quiesce/unquiesce use cases, has been considered neither secure nor generic enough. The controller needs to be granted super user privileges on nodes.
+
+Meanwhile, there are other interesting use cases which requires a similar mechanism:
+* Users have been asking a feature to signal `Pod`s in this [issue](https://github.com/kubernetes/kubernetes/issues/24957). For example, sending a signal to a `Pod` to reload configuration, to flush/roll logs, to change log verbosity etc.
+* Mechanism to notify `Pod`s on a to-be-evicted node.
+
+With these new motivations, we are proposing a more generic design in this KEP. As the eventual goal is to have Kubelet executing the commands against `Pod`s, this approach is considered to be more secure as it will not widen the scope to grant super user privileges.
+
+### Goals
+
+#### Phase 1
+
+- Users can specify an optional list of commands in `Pod`s specification at container level.
+- Users or upper level controllers can request to execute commands in `Pod`s by creation of an API object(`PodNotification`).
+- Implement a *trusted* controller which monitors requests to execute commands, notifies the corresponding `Pod`s to run those commands, and updates status of requests corresponding to the execution results.
+
+#### Phase 2
+
+- Introduce an upper level API(`Notification`) which allows users to send `PodNotification` to a set of `Pod`s which can be matched by a label selector.
+- Extend `ContainerNotifier` API in Phase 1 to support sending signals to containers.
+- Move *trusted* controller logic into Kubelet.
+
+#### Phase 3
+
+- Add a `Probe` to verify the results from the commands/signals if needed.
+
+### Non-Goals
+
+- Writing concrete commands, for example quiesce scripts, which could be executed. It is the responsibility of users who set up pod definitions.
+
+## Proposal
+
+In phase 1, propose to introduce several API changes:
+- Adding an optional field `Notifiers` which is a list of `ContainerNotifier`s into `Container`.
+- Adding an inlined type `ContainerNotifierHandler` which defines a command.
+- Adding a core API type `PodNotification` which defines request type to trigger execution of `ContainerNotifier`(s) in a `Pod`.
```

Comment:
I don’t see a defense/justification for this object in the examples of this KEP (or alternatives) vs simply having a spec generation driven interaction in the pod itself for phase one.  If it’s been discussed and rejected, I’d like it added to alternatived (so i can see the critique).  If it hasn’t been discussed, I’d prefer it to the addition of a new object and controller in phase one that seem like they are “controller for controllers sake”.

Roughly, id expect pod.spec.containers.notifier to have a blank generation field that defaults to spec generation on edit of the notifier struct, or is 0 on create.  And then status to have “notifiers[].generation” that indicates if the kubelet executed the signal.  Clients clear spec generation to trigger the signal and wait for status generation to match to know when the kubelet successfully “leveled” the notifier (delivered it).

That seems a more useful primitive than the hypothetical object and controller, can be done in phase one so external controllers can invoke signals, and doesn’t require the extra alpha baggage until later (bulk notification can be done as a separate orthogonal phase around generation driven signaling).

Reply:
To be honest, this design was trying to minimize the number of moving parts within Pod, but if folks want to explore that as an alternative...

Pod doesn't seem to use `generation` so first we enable that.  Or use RV ?

So to trigger a notifier, we'd write 0 to the pod...notifier generation, which is roughly equivalent to creating a PodNotifier as specced today.  Right?

That does get rid of PodNotifier type in exchange for more moving parts in Pod itself.  In particular if means that any controller which converts a higher-level Notifier (eg, notify all pods under a selector) needs RBAC to mutate Pod.  That seems strictly worse to me, unless we add a subresource for notifications.

Do you perceive that as better?  I don't hate it, and it eliminates a whole category of problems.

We still need a controller.  Two in fact - one to actuate the notifications (eventually kubelet) and one to take higher level "notify by selector" and convert that to N discrete notifications (or we push that up onto all consumers of notifiers).  As you say, that's not phase 1 necessary.